google/export/export_test.go

```diff
@@ -482,54 +437,226 @@ func TestApplyConfig(t *testing.T) {
+				{metric: []string{"ref", "2"}, resource: resourceWithProject("project-test")},
+				{metric: []string{"ref", "3"}, resource: resourceWithProject("project-test")},
 			},
 		},
-	}); err != nil {
-		t.Fatalf("Initial apply: %s", err)
-	}
-	sendAndTestSamples("project-abc")
-
-	// Force new client to fail and ApplyConfig that should NOT create new client.
-	e.newMetricClient = func(_ context.Context, opts ExporterOpts) (metricServiceClient, error) {
-		return nil, errors.New("client should be NOT recreated")
-	}
-	if err := e.ApplyConfig(&config.Config{
-		GlobalConfig: config.GlobalConfig{
-			ExternalLabels: labels.FromStrings(KeyProjectID, "project-xyz"),
+		{
+			name: "Filtering change",
+			toApply: &config.Config{
+				GlobalConfig: config.GlobalConfig{
+					ExternalLabels: labels.FromStrings(KeyProjectID, "project-test"),
+				},
+				GoogleCloud: gcmconfig.GoogleCloudConfig{
+					Export: gcmconfig.GoogleCloudExportConfig{
+						Compression: "gzip",
+						EnableMatch: true,
+						Match:       []string{`{ref=~"(0|3)"}`, `{ref="2"}`}, // Skip 1.
+					},
+				},
+			},
+			expectedSamplesSeries: []gcmSampleLabels{
+				{metric: []string{"ref", "2"}, resource: resourceWithProject("project-test")},
+				{metric: []string{"ref", "3"}, resource: resourceWithProject("project-test")},
+			},
 		},
-		GoogleCloud: gcmconfig.GoogleCloudConfig{
-			Export: gcmconfig.GoogleCloudExportConfig{
-				Compression: "gzip",
+		{
+			name: "Filtering change (noop again)",
+			toApply: &config.Config{
+				GlobalConfig: config.GlobalConfig{
+					ExternalLabels: labels.FromStrings(KeyProjectID, "project-test"),
+				},
+				GoogleCloud: gcmconfig.GoogleCloudConfig{
+					Export: gcmconfig.GoogleCloudExportConfig{
+						Compression: "gzip",
+						EnableMatch: false, // noop.
+					},
+				},
+			},
+			expectedSamplesSeries: []gcmSampleLabels{
+				{metric: []string{"ref", "2"}, resource: resourceWithProject("project-test")},
+				{metric: []string{"ref", "3"}, resource: resourceWithProject("project-test")},
 			},
 		},
-	}); err != nil {
-		t.Fatalf("Initial apply: %s", err)
+		{
+			name: "Filtering change (back)",
+			toApply: &config.Config{
+				GlobalConfig: config.GlobalConfig{
+					ExternalLabels: labels.FromStrings(KeyProjectID, "project-test"),
+				},
+				GoogleCloud: gcmconfig.GoogleCloudConfig{
+					Export: gcmconfig.GoogleCloudExportConfig{
+						Compression: "gzip",
+						EnableMatch: true,
+					},
+				},
+			},
+			expectedCacheEntriesRefreshed: []storage.SeriesRef{1}, // Expect refresh for a series that's now matching.
+			expectedSamplesSeries: []gcmSampleLabels{
+				{metric: []string{"ref", "1"}, resource: resourceWithProject("project-test")},
+				{metric: []string{"ref", "2"}, resource: resourceWithProject("project-test")},
+				{metric: []string{"ref", "3"}, resource: resourceWithProject("project-test")},
+			},
+		},
+	} {
+		t.Run(tcase.name, func(t *testing.T) {
+			t.Helper()
```

Comment:
question: is this actually a test helper function?

It looks like you're doing actual assertions in this scope. Same for L587 and L593.